code/campaignclassic/src/main/java/com/adobe/marketing/mobile/CampaignPushUtils.java

```diff
@@ -35,41 +55,72 @@
-        } finally {
-            if (inputStream != null) {
-                try {
-                    inputStream.close();
-                } catch (IOException e) {
-                    Log.warning(
-                            CampaignPushConstants.LOG_TAG,
-                            SELF_TAG,
-                            "IOException during closing Input stream while push notification image"
-                                    + " from url (%s). Exception: %s ",
-                            url,
-                            e.getMessage());
+        DownloadImageCallable(final String url) {
+            this.url = url;
+        }
+
+        @Override
+        public Bitmap call() {
+            Bitmap bitmap = null;
+            HttpURLConnection connection = null;
+            InputStream inputStream = null;
+
+            try {
+                final URL imageUrl = new URL(url);
+                connection = (HttpURLConnection) imageUrl.openConnection();
+                inputStream = connection.getInputStream();
+                bitmap = BitmapFactory.decodeStream(inputStream);
+            } catch (final IOException e) {
+                Log.warning(
+                        CampaignPushConstants.LOG_TAG,
+                        SELF_TAG,
+                        "Failed to download push notification image from url (%s). Exception: %s",
+                        url,
+                        e.getMessage());
+            } finally {
+                if (inputStream != null) {
+                    try {
+                        inputStream.close();
+                    } catch (final IOException e) {
+                        Log.warning(
+                                CampaignPushConstants.LOG_TAG,
+                                SELF_TAG,
+                                "IOException during closing Input stream while push notification"
+                                        + " image from url (%s). Exception: %s ",
+                                url,
+                                e.getMessage());
+                    }
                 }
-            }
 
-            if (connection != null) {
-                connection.disconnect();
+                if (connection != null) {
+                    connection.disconnect();
+                }
             }
+
+            Log.trace(
+                    CampaignPushConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Downloaded push notification image from url (%s)",
+                    url);
+            return bitmap;
+        }
+    }
+    ;
+
+    static Bitmap download(final String url) {
+        Bitmap bitmap = null;
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        final Future<Bitmap> downloadTask = executorService.submit(new DownloadImageCallable(url));
+
+        try {
+            bitmap = downloadTask.get(1, TimeUnit.SECONDS);
+            executorService.shutdown();
```

Comment:
How long does it typically taking for downloads? Default Core NetworkService read and connection timeout are 10seconds. So it seems to me that 1second is too aggressive. Does Firebase require us to handle the message in a set time?

Reply:
i'll increase this to 10 seconds as well. Firebase expects the message to be handled within 20 seconds so this will be within that timeframe.